balancer/balanceraggregator/balanceraggregator.go

```diff
@@ -0,0 +1,265 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package balanceraggregator implements a BalancerAggregator helper.
+package balanceraggregator
+
+import (
+	"encoding/json"
+	"errors"
+	"sync/atomic"
+
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/balancer/base"
+	"google.golang.org/grpc/connectivity"
+	"google.golang.org/grpc/internal/balancer/gracefulswitch"
+	"google.golang.org/grpc/internal/grpcrand"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/serviceconfig"
+)
+
+// ChildState is the balancer state of the child along with the
+// endpoint which ID's the child balancer.
+type ChildState struct {
+	Endpoint resolver.Endpoint
+	State    balancer.State
+}
+
+// Build returns a new BalancerAggregator.
+func Build(cc balancer.ClientConn, opts balancer.BuildOptions) *BalancerAggregator {
+	return &BalancerAggregator{
+		cc:       cc,
+		bOpts:    opts,
+		children: resolver.NewEndpointMap(),
+	}
+}
+
+// BalancerAggregator is a balancer that wraps child balancers. It creates a
+// child balancer with child config for every Endpoint received. It updates the
+// child states on any update from parent or child.
+type BalancerAggregator struct {
+	cc    balancer.ClientConn
+	bOpts balancer.BuildOptions
+
+	children *resolver.EndpointMap
+
+	inhibitChildUpdates bool
+}
+
+// UpdateClientConnState creates a child for new endpoints, deletes children for
+// endpoints that are no longer present. It also updates all the children, and
+// sends a single synchronous update of the children's aggregate state at the
+// end of the UpdateClientConnState operation. If an endpoint has no addresses,
+// returns error. Otherwise returns first error found from a child, but fully
+// processes the new update.
+func (ba *BalancerAggregator) UpdateClientConnState(state balancer.ClientConnState) error {
+	if len(state.ResolverState.Endpoints) == 0 {
+		return errors.New("endpoints list is empty")
+	}
+
+	// Check/return early if any endpoints have no addresses.
+	// TODO: eventually make this configurable.
+	for _, endpoint := range state.ResolverState.Endpoints {
+		if len(endpoint.Addresses) == 0 {
+			return errors.New("endpoint has empty addresses")
+		}
+	}
+	endpointSet := resolver.NewEndpointMap()
+	ba.inhibitChildUpdates = true
+	defer func() {
+		ba.inhibitChildUpdates = false
+		ba.updateState()
+	}()
+	var ret error
+	// Update/Create new children.
+	for _, endpoint := range state.ResolverState.Endpoints {
+		endpointSet.Set(endpoint, nil)
```

Comment:
Hmm good point...although thinking about this, you could get a endpoint {a, b, c} with attribute x, and endpoint {b, c, a} with attribute y...what do we want the behavior to be here (edge case that I don't think should come out of any of our maintained resolvers). The code as written uses the last positional endpoint to determine the attributes. This change makes it use the first positional endpoint to determine the attributes.

Reply:
I changed it to what you had to avoid duplicate updates, but question here about attributes still stands.

Reply:
This is what the EndpointMap is for, right?  If they have different attributes but the same addresses, then they should be treated as unique.  (This applies for address attributes at least.  Not sure about endpoint attributes, but that seems reasonable too, since it could affect the behavior of the LB policy that handles them.)